src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -0,0 +1,749 @@
+import is.hail.expr.types._
+import is.hail.utils._
+import org.apache.spark.sql.Row
+
+object PruneDeadFields {
+  def isSupertype(superType: BaseType, subType: BaseType): Boolean = {
+    try {
+      (superType, subType) match {
+        case (tt1: TableType, tt2: TableType) =>
+          isSupertype(tt1.globalType, tt2.globalType) &&
+            isSupertype(tt1.rowType, tt2.rowType)
+        case (mt1: MatrixType, mt2: MatrixType) =>
+          isSupertype(mt1.globalType, mt2.globalType) &&
+            isSupertype(mt1.rowType, mt2.rowType) &&
+            isSupertype(mt1.colType, mt2.colType) &&
+            isSupertype(mt1.entryType, mt2.entryType)
+        case (TArray(et1, r1), TArray(et2, r2)) => r1 == r2 && isSupertype(et1, et2)
+        case (TSet(et1, r1), TSet(et2, r2)) => r1 == r2 && isSupertype(et1, et2)
+        case (TDict(kt1, vt1, r1), TDict(kt2, vt2, r2)) => r1 == r2 && isSupertype(kt1, kt2) && isSupertype(vt1, vt2)
+        case (s1: TStruct, s2: TStruct) =>
+          var idx = -1
+          s1.required == s2.required && s1.fields.forall { f =>
+            val s2field = s2.field(f.name)
+            if (s2field.index > idx) {
+              idx = s2field.index
+              isSupertype(f.typ, s2field.typ)
+            } else
+              false
+          }
+        case (agg1: TAggregable, agg2: TAggregable) => isSupertype(agg1.elementType, agg2.elementType)
+        case (t1: Type, t2: Type) => t1 == t2
+        case _ => fatal(s"invalid comparison: $superType / $subType")
+      }
+    } catch {
+      case e: Throwable =>
+        fatal(s"error while checking subtype:\n  super: $superType\n  sub:   $subType", e)
+    }
+  }
+
+  def apply(ir: BaseIR): BaseIR = {
+
+    try {
+      ir.deepCopy() match {
+        case x: MatrixIR => rebuild(x, x.typ)
+        case x: TableIR => rebuild(x, x.typ)
+        case x: IR => rebuild(x)
+      }
+    } catch {
+      case e: Throwable => fatal(s"error trying to rebuild IR:\n${ Pretty(ir) }", e)
+    }
+  }
+
+
+  def minimal(tt: TableType): TableType = {
+    val keySet = tt.key.iterator.flatten.toSet
+    tt.copy(
+      rowType = tt.rowType.filter(keySet)._1,
+      globalType = TStruct(tt.globalType.required)
+    )
+  }
+
+  def minimal(mt: MatrixType): MatrixType = {
+    val rowKeySet = mt.rowKey.toSet
+    val colKeySet = mt.colKey.toSet
+    mt.copy(
+      rvRowType = mt.rvRowType.filter(rowKeySet ++ Array(MatrixType.entriesIdentifier))._1,
+      colType = mt.colType.filter(colKeySet)._1,
+      globalType = TStruct(mt.globalType.required)
+    )
+  }
+
+  def minimal[T <: Type](base: T): T = {
+    val result = base match {
+      case ts: TStruct => TStruct(ts.required)
+      case ta: TArray => TArray(minimal(ta.elementType), ta.required)
+      case tagg: TAggregable => tagg.copy(elementType = minimal(tagg.elementType))
+      case t => t
+    }
+    result.asInstanceOf[T]
+  }
+
+  def unionBaseType[T <: BaseType](base: T, children: T*): T = {
```

Comment:
And the same idiom here.

Reply:
In particular, since this is taking a BaseType, it should handle all the (base) types, including types.